File: src/java/4_BehavioralPatterns/1_ChainOfResponsibility/ChainOfResponsibilityExercise.java
package com.activemesa.behavioral.chainofresponsibility.exercise;

import java.util.ArrayList;
import java.util.List;

abstract class Creature
{
  protected Game game;
  protected int baseAttack, baseDefense;

  public Creature(Game game, int baseAttack, int baseDefense)
  {
    this.game = game;
    this.baseAttack = baseAttack;
    this.baseDefense = baseDefense;
  }

  public abstract int getAttack();
  public abstract int getDefense();
  public abstract void query(Object source, StatQuery sq);
}

class Goblin extends Creature
{
  protected Goblin(Game game, int baseAttack, int baseDefense)
  {
    super(game, baseAttack, baseDefense);
  }

  public Goblin(Game game)
  {
    super(game, 1, 1);
  }

  @Override
  public int getAttack()
  {
    StatQuery q = new StatQuery(Statistic.ATTACK);
    for (Creature c : game.creatures)
      c.query(this, q);
    return q.result;
  }

  @Override
  public int getDefense()
  {
    StatQuery q = new StatQuery(Statistic.DEFENSE);
    for (Creature c : game.creatures)
      c.query(this, q);
    return q.result;
  }

  @Override
  public void query(Object source, StatQuery sq)
  {
    if (source == this)
    {
      switch (sq.statistic)
      {
        case ATTACK:
          sq.result += baseAttack;
          break;
        case DEFENSE:
          sq.result += baseDefense;
          break;
      }
    }
    else if (sq.statistic == Statistic.DEFENSE)
    {
      sq.result++;
    }
  }
}

class GoblinKing extends Goblin
{
  public GoblinKing(Game game)
  {
    super(game, 3, 3);
  }

  @Override
  public void query(Object source, StatQuery sq)
  {
    if (source != this && sq.statistic == Statistic.ATTACK)
    {
      sq.result++; // every goblin gets +1 attack
    }
    else super.query(source, sq);
  }
}

enum Statistic
{
  ATTACK, DEFENSE
}

class StatQuery
{
  public Statistic statistic;
  public int result;

  public StatQuery(Statistic statistic)
  {
    this.statistic = statistic;
  }
}

class Game
{
  public List<Creature> creatures = new ArrayList<>();
}

File: src/js/4_BehavioralPatterns/1_ChainOfResponsibility/chain_of_responsibility_exercise.js
let WhatToQuery = Object.freeze({
  'attack': 1,
  'defense': 2
});

class Query
{
  constructor(whatToQuery, result)
  {
    this.whatToQuery = whatToQuery;
    this.result = result;
  }
}

class Goblin
{
  constructor(game, baseAttack=1, baseDefense=1)
  {
    this.game = game;
    game.creatures.push(this);
    this.baseAttack = baseAttack;
    this.baseDefense = baseDefense;
  }

  handleQuery(source, query)
  {
    if (source === this)
    {
      switch (query.whatToQuery)
      {
        case WhatToQuery.attack:
          query.result += this.baseAttack;
          break;
        case WhatToQuery.defense:
          query.result += this.baseDefense;
          break;
      }
    }
    else if (query.whatToQuery === WhatToQuery.defense)
    {
      query.result++;
    }
  }

  get defense()
  {
    let q = new Query(WhatToQuery.defense, 0);
    for (let c of this.game.creatures)
      c.handleQuery(this, q);
    return q.result;
  }

  get attack()
  {
    let q = new Query(WhatToQuery.attack, 0);
    for (let c of this.game.creatures)
      c.handleQuery(this, q);
    return q.result;
  }
}

class GoblinKing extends Goblin
{
  constructor(game)
  {
    super(game, 3, 3);
  }

  handleQuery(source, query) {
    if (source !== this && query.whatToQuery === WhatToQuery.attack)
    {
      query.result++;
    }
    super.handleQuery(source, query);
  }
}

class Game
{
  constructor()
  {
    this.creatures = [];
  }
}

describe('chain of responsibility', function()
{
  it('should handle many goblins', function()
  {
    let game = new Game();
    let goblin = new Goblin(game);

    expect(goblin.attack).toEqual(1);
    expect(goblin.defense).toEqual(1);

    let goblin2 = new Goblin(game);

    expect(goblin.attack).toEqual(1);
    expect(goblin.defense).toEqual(2);

    let goblin3 = new GoblinKing(game);

    expect(goblin.attack).toEqual(2);
    expect(goblin.defense).toEqual(3);
  });
});